File: core/src/com/dpashko/sandbox/model/ModelProvider.java
package com.dpashko.sandbox.model;

import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g3d.Material;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.badlogic.gdx.graphics.g3d.loader.ObjLoader;
import com.dpashko.sandbox.files.FileProvider;
import com.dpashko.sandbox.material.MaterialProvider;
import org.jetbrains.annotations.NotNull;

public class ModelProvider {

    private static final ObjLoader objLoader = new ObjLoader();
    private static ModelInstance cachedBoxModel = null;
    private static ModelInstance cachedCylinderModel = null;
    private static ModelInstance cachedSphereModel = null;

    @NotNull
    public static ModelInstance createSphere() {
        if (cachedSphereModel == null) {
            cachedSphereModel = loadModel(FileProvider.sphereModel, null);
        }
        return cachedSphereModel.copy();
    }

    @NotNull
    public static ModelInstance createBox() {
        return createBox(1f);
    }

    @NotNull
    public static ModelInstance createBox(float size) {
        if (cachedBoxModel == null) {
            cachedBoxModel = loadModel(FileProvider.boxModel, MaterialProvider.diffuse(Color.GRAY));
        }
        ModelInstance modelInstance = cachedBoxModel.copy();
        modelInstance.transform.scl(size);
        return modelInstance;
    }

    @NotNull
    public static ModelInstance createCylinder() {
        if (cachedCylinderModel == null) {
            cachedCylinderModel = loadModel(FileProvider.cylinderModel, null);
        }
        return cachedCylinderModel.copy();
    }


    @NotNull
    public static ModelInstance loadModel(final FileHandle file, final Material material) {
        final Model model = objLoader.loadModel(file, true);
        model.materials.clear();
        if (material != null)
            model.materials.add(material);
        return new ModelInstance(model);
    }
}

File: core/src/com/dpashko/sandbox/model/ModelProvider.kt
package com.dpashko.sandbox.model

import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g3d.Material
import com.badlogic.gdx.graphics.g3d.ModelInstance
import com.badlogic.gdx.graphics.g3d.loader.ObjLoader
import com.dpashko.sandbox.files.FileProvider
import com.dpashko.sandbox.material.MaterialProvider

object ModelProvider {

    private val objLoader = ObjLoader()

    @JvmOverloads
    fun createSphere(
        material: Material = MaterialProvider.diffuse(Color.GRAY)
    ) = loadModel(FileProvider.sphereModel, material)

    @JvmOverloads
    fun createCylinder(
        material:
        Material = MaterialProvider.diffuse(Color.GRAY)
    ) = loadModel(FileProvider.cylinderModel, material)

    @JvmOverloads
    fun createBox(
        material: Material = MaterialProvider.diffuse(Color.GRAY)
    ) = loadModel(FileProvider.boxModel, material)

    @JvmOverloads
    fun createSuzanne(
        material: Material = MaterialProvider.diffuse(Color.GRAY)
    ) = loadModel(FileProvider.suzanneModel, material)

    @JvmOverloads
    fun createActor(
        material: Material = MaterialProvider.diffuse(Color.GRAY)
    ) = loadModel(FileProvider.actor, material)

    fun createPlayground() = loadModel(FileProvider.playground, null)

    public fun loadModel(file: FileHandle?, material: Material?): ModelInstance {
        val model = objLoader.loadModel(file, true)
        if (material != null) {
            model.materials.clear()
            model.materials.add(material)
        }
        return ModelInstance(model)
    }
}
